File: lib/services/expense_service.dart
import 'dart:async';
import '../models/expense.dart';
import 'firebase_service.dart';

class ExpenseService {
  final FirebaseService _firebase;
  final String _path = 'expenses';

  ExpenseService(this._firebase);

  Future<String> create(Expense expense) async {
    return await _firebase.save(_path, expense.toJson());
  }

  Future<void> update(Expense expense) async {
    if (expense.id == null) throw Exception('Expense ID is required');
    await _firebase.update(_path, expense.id!, expense.toJson());
  }

  Future<void> delete(String id) async {
    await _firebase.delete(_path, id);
  }

  Future<Expense?> get(String id) async {
    final data = await _firebase.get(_path, id);
    return data != null ? Expense.fromJson(data, id) : null;
  }

  Future<List<Expense>> getAll() async {
    final data = await _firebase.getAll(_path);
    return data.map((item) => Expense.fromJson(item, item['id'])).toList();
  }

  Future<List<Expense>> getByDateRange(DateTime start, DateTime end) async {
    final expenses = await getAll();
    return expenses.where((expense) {
      return expense.date.isAfter(start.subtract(const Duration(days: 1))) &&
          expense.date.isBefore(end.add(const Duration(days: 1)));
    }).toList();
  }

  Future<double> getTotal(DateTime start, DateTime end) async {
    final expenses = await getByDateRange(start, end);
    return expenses.fold<double>(0.0, (double sum, Expense expense) => sum + expense.amount);
  }
}

File: js/services/expense-service.js
import { database } from '../firebase-config.js';
import { ref, push, set, get, remove } from 'firebase/database';

export class ExpenseService {
    constructor() {
        this.path = 'expenses';
    }

    async create(expense) {
        const expenseRef = ref(database, this.path);
        const newRef = push(expenseRef);
        await set(newRef, expense);
        return newRef.key;
    }

    async update(id, expense) {
        const expenseRef = ref(database, `${this.path}/${id}`);
        await set(expenseRef, expense);
    }

    async delete(id) {
        const expenseRef = ref(database, `${this.path}/${id}`);
        await remove(expenseRef);
    }

    async get(id) {
        const expenseRef = ref(database, `${this.path}/${id}`);
        const snapshot = await get(expenseRef);
        if (snapshot.exists()) {
            return { id: snapshot.key, ...snapshot.val() };
        }
        return null;
    }

    async getAll() {
        const expensesRef = ref(database, this.path);
        const snapshot = await get(expensesRef);
        const expenses = [];
        if (snapshot.exists()) {
            snapshot.forEach(child => {
                expenses.push({ id: child.key, ...child.val() });
            });
        }
        return expenses;
    }

    async getByDateRange(startDate, endDate) {
        const expenses = await this.getAll();
        return expenses.filter(expense => {
            const expenseDate = new Date(expense.date);
            return expenseDate >= startDate && expenseDate <= endDate;
        });
    }

    async getTotal(startDate, endDate) {
        const expenses = await this.getByDateRange(startDate, endDate);
        return expenses.reduce((sum, expense) => sum + (expense.amount || 0), 0);
    }
}

export const expenseService = new ExpenseService();




